### DISCUSSION TITLE
Cache management when updating a multi-instance production

### DISCUSSION THREAD

**Discussion by CorentinDeBoisset** [symfony/symfony#42646] [Q&A]
Hello,

I am preparing to shift our production into a multi-instance system, based on docker and operated by something in the likes of kubernetes or nomad. The question of the deployment is one that is starting to get increasingly complex, and there are still questions that I don't know how to handle:

When starting a rolling update (or a canary/blue-green), how can one manage the symfony system cache ? If all the instance share the same cache server, warming up the cache in an instance with updated code will poison the cache of the previous, still-running containers. Conversely, blocking a cache warm-up until all the new instances are spin up means they will not be functionnal until then.

The DevOps solution for this kind of issue for database migrations is to ensure the migrations are always backward-compatible, meaning both the old and new instances can run with the updated schema, however I don't think this kind of method exists for the symfony system cache?

Would a solution be to keep the system cache pool in the filesystem, locally and independent for every container, and only move the other cache pools in a shared memcached/redis/pdo/... ?

Thanks !

**Comment by romaricdrigon** [comment]
What do you mean exactly by "cache"?
If it is Symfony cache, the best option IMO is to have each instance have its own cache*, in the container, not mounted to any volume. It will bring better IO performances, and you can even build the cache into the container image (Symfony 5 is designed to have a read-only cache ; a few bundles may not respect this, though, so to test before).

Regarding userland caching, it depends upon what you use your cache for., so we will need more information on that.

* either `var/cache/` folder, either in its memory (if using APCu...).
